### ISSUE TITLE
[ide] macro method documentation is taken from macro implementation rather than entry-point definition

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#10234]
Main.hx:
```haxe
/** proper documentation **/
macro function macroFunc();

function main() {
	macroFunc();
}
```

Main.macro.hx:
```haxe
/** it gets documentation from here **/
function macroFunc() return macro {};
```

Hovering (or autocompleting) `macroFunc` shows documentation from the `.macro.hx` instead of "proper documentation" from the entry-point definition, which is a bit unexpected from the user's standpoint.

![image](https://user-images.githubusercontent.com/49749/116375697-d109fa80-a80f-11eb-9893-2059a20a777c.png)

I guess the question is whether this is something we have to live with (I assume it's consistent with the general `.platform.hx` behaviour?) or we could change this to follow the principle of least surprise? :)

**Comment by nadako** [comment]
BTW, the IDE services also don't show ExprOf signatures in this case, looks like it fully ignores the entry point definition:
![image](https://user-images.githubusercontent.com/49749/116376430-889f0c80-a810-11eb-8d7f-42ad645373dc.png)

**Comment by nadako** [comment]
One more related thing, hovering an "instance" macro method doesn't skip the first argument:

![image](https://user-images.githubusercontent.com/49749/116377000-0f53e980-a811-11eb-97c8-9028279f678d.png)

**Comment by Aurel300** [comment]
> which is a bit unexpected from the user's standpoint.

I am not sure I agree with this. I would be a lot more surprised if everything about the function (its implementation, its actual signature) is taken from the `.macro.hx` but its documentation somehow isn't.

**Comment by Simn** [comment]
Yeah this seems somewhat dodgy. You could make the case that the non-macro documentation should be taken as a fallback if the macro version doesn't have one, but otherwise the current behavior is more consistent.

**Comment by back2dos** [comment]
Let's consider this example (with an instance macro with a conflicting signature for maximum mayhem):

```haxe
class Main {
  function new() {}
  static function main() {
    trace(new Main().greeting());
  }
  /**
   * Haha!!!
   */
  macro function greeting(a, b, c);
}

// .macro.hx
class Main {
  static function greeting(_)
    return macro 'hello world';
}
```

Let's call the `macro` function "declaration" (because it is purely declarative) and the static function from `.macro.hx` the "implementation". One could argue that the only thing being declared is the presence of an implementation (which also defines the signature and therefore should be the authoritative source of documentation). This poses a problem however, because in the non-macro context, the `KMacro` method is all that exists and as a result:

1. dox picks up the declaration - in the absence of `-xml` one would use `onGenerate`, which also sees the declaration
2. Auto completion shows the declaration
3. `Goto Definition` also jumps to the declaration too. 

When hovering over `.greeting()` you get the implementation, when `Ctrl`-hovering, you get both (kinda neat) and when clicking, you jump to the declaration (which is not what you saw when first hovering). 

Also, I noticed that `a` is stripped from the argument list, because it's an instance macro. This can also be observed in the signature hint:

![image](https://user-images.githubusercontent.com/509501/116526021-7046f480-a8d9-11eb-94f1-ee3816c94d8b.png)

That can be seen as a separate issue, but I think the underlying problem is that the way in which the declaration and implementation are combined grew in a somewhat ad-hoc fashion and should be reviewed altogether to achieve the abovementioned consistency.

Personally, I like the idea of the declaration being the signature (and thus also the source of documentation). It would also allow using plain `Int` instead of `ExprOf<Int>` and would give the user a nice concise signature altogether. That being said, there's also some merit to the compiler completely ignoring the declaration's signature, because you then don't have to spell things out twice. I have actually relied on this in a few places, but from my end I wouldn't mind having that broken.

**Comment by frabbit** [comment]
It seems that there is also no validation if the same methods exists in macro context. Consider a typo like "foobar" vs "fobar". The typo will only surface if a macro call expression exists and the compiler checks for its existance.
